File: native-ios/RideTrackerNative/CalibrationMath.swift
import Foundation
import simd

enum ForwardEdge: String, CaseIterable, Codable, Identifiable {
    case top, bottom, left, right
    var id: String { rawValue }
    var title: String {
        switch self {
        case .top: return "Oberkante"
        case .bottom: return "Unterkante"
        case .left: return "Linke Kante"
        case .right: return "Rechte Kante"
        }
    }
    var vector: SIMD3<Double> {
        switch self {
        case .top: return SIMD3(0, -1, 0)
        case .bottom: return SIMD3(0, 1, 0)
        case .left: return SIMD3(-1, 0, 0)
        case .right: return SIMD3(1, 0, 0)
        }
    }
}

enum CalibrationMath {
    static func build(gravitySamples: [SIMD3<Double>], forwardEdge: ForwardEdge) -> SeatCalibration? {
        guard gravitySamples.count >= 20 else { return nil }
        let sum = gravitySamples.reduce(SIMD3<Double>(repeating: 0), +)
        let average = sum / Double(gravitySamples.count)
        guard simd_length(average) > 0.5 else { return nil }
        let up = simd_normalize(average)
        let hint = forwardEdge.vector
        var projected = hint - simd_dot(hint, up) * up
        if simd_length(projected) < 0.2 {
            let fallback = abs(up.x) < 0.8 ? SIMD3<Double>(1, 0, 0) : SIMD3<Double>(0, 1, 0)
            projected = simd_cross(up, fallback)
        }
        let forward = simd_normalize(projected)
        let lateral = simd_normalize(simd_cross(forward, up))
        let correctedForward = simd_normalize(simd_cross(up, lateral))
        return SeatCalibration(up: up, lateral: lateral, forward: correctedForward, source: "manual")
    }
}

File: native-android/app/src/main/java/de/ridetracker/engine/CalibrationMath.kt
package de.ridetracker.engine

import kotlin.math.abs
import kotlin.math.sqrt

enum class ForwardEdge(val title: String, val vector: Vector3) {
    TOP("Oberkante", Vector3(0.0, -1.0, 0.0)),
    BOTTOM("Unterkante", Vector3(0.0, 1.0, 0.0)),
    LEFT("Linke Kante", Vector3(-1.0, 0.0, 0.0)),
    RIGHT("Rechte Kante", Vector3(1.0, 0.0, 0.0)),
}

object CalibrationMath {
    fun build(samples: List<Vector3>, edge: ForwardEdge): SeatCalibration? {
        if (samples.size < 20) return null
        val avg = samples.reduce(::add).let { scale(it, 1.0 / samples.size) }
        if (length(avg) < 0.5) return null
        val up = normalize(avg)
        val hint = edge.vector
        var projected = subtract(hint, scale(up, dot(hint, up)))
        if (length(projected) < 0.2) {
            val fallback = if (abs(up.x) < 0.8) Vector3(1.0, 0.0, 0.0) else Vector3(0.0, 1.0, 0.0)
            projected = cross(up, fallback)
        }
        val forward = normalize(projected)
        val lateral = normalize(cross(forward, up))
        val correctedForward = normalize(cross(up, lateral))
        return SeatCalibration(up, lateral, correctedForward, "manual")
    }

    private fun add(a: Vector3, b: Vector3) = Vector3(a.x + b.x, a.y + b.y, a.z + b.z)
    private fun subtract(a: Vector3, b: Vector3) = Vector3(a.x - b.x, a.y - b.y, a.z - b.z)
    private fun scale(v: Vector3, s: Double) = Vector3(v.x * s, v.y * s, v.z * s)
    private fun dot(a: Vector3, b: Vector3) = a.x * b.x + a.y * b.y + a.z * b.z
    private fun cross(a: Vector3, b: Vector3) = Vector3(a.y*b.z-a.z*b.y, a.z*b.x-a.x*b.z, a.x*b.y-a.y*b.x)
    private fun length(v: Vector3) = sqrt(dot(v, v))
    private fun normalize(v: Vector3): Vector3 { val l = length(v).coerceAtLeast(1e-9); return scale(v, 1.0/l) }
}
